api/src/main/java/org/openmrs/module/poc/clinicalservice/service/ClinicalServiceServiceImpl.java

```diff
@@ -152,36 +154,93 @@ private void deleteOldObsOfClinicalService(final Encounter fetchedEncounter, fin
 		        .getClinicalServices(ClinicalServiceKeys.getClinicalServiceByCode(serviceCode));
 		
 		final Set<Obs> allOldObs = fetchedEncounter.getAllObs(false);
-		for (final Obs oldItemObs : allOldObs) {
+		final List<Obs> obsGroups = new ArrayList<>();
+		for (final Obs oldObsItem : allOldObs) {
 			
-			final String conceptUuid = oldItemObs.getConcept().getUuid();
-			if ((oldItemObs.getObsId() != null) && clinicalServices.contains(conceptUuid)
+			final String conceptUuid = oldObsItem.getConcept().getUuid();
+			if ((oldObsItem.getObsId() != null) && clinicalServices.contains(conceptUuid)
 			        && !ClinicalServiceServiceImpl.EXCLUDING_CONCEPTS.contains(conceptUuid)) {
-				this.obsService.voidObs(oldItemObs,
-				    Context.getMessageSourceService().getMessage("poc.global.info.voided.forupdate"));
+				this.voidObs(oldObsItem);
+				
+				if ((oldObsItem.getObsGroup() != null) && !obsGroups.contains(oldObsItem.getObsGroup())) {
```

Comment:
Would using a `Set` be more appropriate than using `List` and checking if the item is already contined?

Reply:
Sure. Before I used Set with implementation of TreeSet and was getting ClasssCastException once Obs are not Comparable. That's why I Used list.

Will change to use Set with implementatio of LinkedHashSet or simply HashSet , these two implementation also doesn't  allow duplicates,